File: 2024/day22.ml
let mix = Int.logxor
let prune n = n mod 16777216

let evolve a =
  let step1 = a * 64 |> mix a |> prune in
  let step2 = step1 / 32 |> mix step1 |> prune in
  step2 * 2048 |> mix step2 |> prune

let nevolve n a =
  Seq.iterate evolve a |> Seq.drop n |> Seq.take 1 |> List.of_seq |> List.hd

let price n = n mod 10

type diffs = int * int * int * int

let npricediffs n a =
  let prices = Seq.iterate evolve a |> Seq.take (n + 1) |> Seq.map price in
  let drop1 = Seq.drop 1 prices in
  let pricediffs = Seq.map2 (fun a b -> b - a) prices drop1 in
  let pricediffs4 =
    Seq.zip pricediffs (Seq.drop 1 pricediffs)
    |> Seq.zip (Seq.drop 2 pricediffs)
    |> Seq.zip (Seq.drop 3 pricediffs)
  in
  Seq.zip pricediffs4 (Seq.drop 4 prices)
  |> Seq.map (fun ((a, (b, (c, d))), p) -> ((c, d, b, a), p))
  |> List.of_seq

module DiffsMap = Map.Make (struct
  type t = diffs

  let compare = compare
end)

let max_diffs m =
  DiffsMap.fold
    (fun d v (dacc, vacc) ->
      let m = max vacc v in
      if m == v then (d, v) else (dacc, vacc))
    m
    ((0, 0, 0, 0), 0)

let find_max_diffs (diffslists : (diffs * int) list list) =
  let combined =
    List.map (fun dl -> List.rev dl |> DiffsMap.of_list) diffslists
    |> List.fold_left
         (DiffsMap.union (fun k v1 v2 -> Some (v1 + v2)))
         DiffsMap.empty
  in
  max_diffs combined

let printkv ((a, b, c, d), value) =
  Printf.printf "%d (%d, %d, %d, %d)\n" value a b c d

let () =
  let file = open_in "/Users/tarunbod/Documents/aoc/2024/inputs/day22.txt" in
  let secrets = In_channel.input_lines file |> List.map int_of_string in
  secrets
  |> List.map (nevolve 2000)
  |> List.fold_left ( + ) 0 |> Printf.printf "%d\n";

  secrets |> List.map (npricediffs 2000) |> find_max_diffs |> printkv

File: 2024/day22.hs
import Data.Bits (xor)
import Data.List (zip4)
import qualified Data.Map as Map

prune :: Int -> Int
prune n = n `mod` 16777216

evolve a =
  let step1 = prune $ xor a $ a * 64
      step2 = prune $ xor step1 $ step1 `div` 32
   in prune $ xor step2 $ step2 * 2048

nEvolve :: Int -> Int -> Int
nEvolve n a = iterate evolve a !! n

price :: Int -> Int
price n = n `mod` 10

type DiffSeq = (Int, Int, Int, Int)

nPriceDiffs :: Int -> Int -> [(DiffSeq, Int)]
nPriceDiffs n a = do
  let prices = map price $ take (n + 1) $ iterate evolve a
  let pricediffs = zipWith (flip (-)) prices $ drop 1 prices
  let pricediffs4 = zip4 pricediffs (drop 1 pricediffs) (drop 2 pricediffs) (drop 3 pricediffs)
  zip pricediffs4 (drop 4 prices)

findMaxDiffs :: [[(DiffSeq, Int)]] -> Int
findMaxDiffs dl = do
  let allMaps = map (Map.fromList . reverse) dl
  let combinedMap = foldl1 (Map.unionWith (+)) allMaps
  maximum $ map snd $ Map.toList combinedMap

main = do
  contents <- getContents
  let nums = map (read :: String -> Int) $ lines contents
  print $ sum $ map (nEvolve 2000) nums
  print $ findMaxDiffs $ map (nPriceDiffs 2000) nums
